lib/src/upstream.rs

```diff
@@ -164,21 +168,47 @@ pub fn send_request(
         backend,
     );
 
+    let try_decompression = req
+        .extensions()
+        .get::<ViceroyRequestMetadata>()
+        .map(|vrm| {
+            vrm.auto_decompress_encodings
+                .contains(ContentEncodings::GZIP)
+        })
+        .unwrap_or(false);
+
     filter_outgoing_headers(req.headers_mut());
     req.headers_mut().insert(hyper::header::HOST, host);
     *req.uri_mut() = uri;
 
     async move {
-        Ok(Client::builder()
+        let basic_response = Client::builder()
             .set_host(false)
             .build(connector)
             .request(req)
             .await
             .map_err(|e| {
                 eprintln!("Error: {:?}", e);
                 e
-            })?
-            .map(Body::from))
+            })?;
+
+        if try_decompression
+            && basic_response.headers().get(header::CONTENT_ENCODING)
+                == Some(&HeaderValue::from_static("gzip"))
```

Comment:
We should also allow for `x-gzip` per https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#gzip.coding (and the C@E implementation).